File: node/routes/repo.js
import express from "express";
import { processRepo } from "../services/process.js";
import { nanoid } from "nanoid";
import { createClient } from "@supabase/supabase-js";

const route = express.Router();
const supabase = createClient(process.env.SUPABASE_URL, process.env.SUPABASE_KEY);

route.post('/ingest', async (req, res) => {
    try {
        const { repoUrl } = req.body;
        if (!repoUrl) {
            return res.status(400).json({ error: "Repository URL is required" })
        }
        const repoId = nanoid(10);
        const repoName = repoUrl
            .split("/")
            .filter(Boolean)
            .pop()
            .replace(".git", "");
        
        await supabase.from("repos").insert({
            id: repoId,
            name: repoName,
            status: "processing"
        })
        res.json({ status: "processing", repoId })
        processRepo(repoUrl, repoId);
    } catch (error) {
        console.log(error);
        res.status(500).json({ error: "Internal server error" })
    }
})

route.get('/status/:repoId', async (req, res) => {
    try {
        const { repoId } = req.params;
        const { data, error } = await supabase.from("repos").select("status").eq("id", repoId);
        if (error) {
            return res.status(500).json({ error: "Internal server error" })
        }
        res.json(data);
    } catch (error) {
        console.log(error);
        res.status(500).json({ error: "Internal server error" })
    }
})
export default route;

File: python/routes/repo.py
from fastapi import APIRouter, BackgroundTasks
from nanoid import generate
from supabase import create_client
from dotenv import load_dotenv
from services.process import process_repo
import os

load_dotenv()

router = APIRouter()

supabase = create_client(os.getenv("SUPABASE_URL"), os.getenv("SUPABASE_KEY"))

@router.post("/ingest")
def ingest_repo(body: dict, background_tasks: BackgroundTasks):
    repo_url = body.get("repo_url")
    if not repo_url:
        return {"error": "Repository URL is required"}
    
    repo_id = generate(size=10)
    repo_name = repo_url.split("/").pop().replace(".git", "")
    response = (
        supabase
        .table("repos")
        .insert({
            "id": repo_id,
            "name": repo_name,
            "status": "processing"
        })
        .execute()
    )

    background_tasks.add_task(process_repo, repo_url, repo_id)
    return {"status": "processing", "repo_id": repo_id}